core/controllers/admin.py

```diff
@@ -854,3 +854,48 @@ def post(self):
             self.render_json({})
         else:
             raise self.InvalidInputException('This app cannot send emails.')
+
+
+class UpdateUsernameHandler(base.BaseHandler):
+    """Handler for renaming usernames."""
+
+    @acl_decorators.can_access_admin_page
+    def put(self):
+        old_username = self.payload.get('old_username', None)
+        new_username = self.payload.get('new_username', None)
+
+        if old_username is None:
+            raise self.InvalidInputException(
+                'Invalid request: The old username must be specified.')
+
+        if new_username is None:
+            raise self.InvalidInputException(
+                'Invalid request: A new username must be specified.')
+
+        if not isinstance(old_username, python_utils.UNICODE):
+            raise self.InvalidInputException(
+                'Expected old username to be a unicode string, received %s'
+                % old_username)
+
+        if not isinstance(new_username, python_utils.UNICODE):
+            raise self.InvalidInputException(
+                'Expected new username to be a unicode string, received %s'
+                % new_username)
+
+        user_id = user_services.get_user_id_from_username(old_username)
+        if user_id is None:
+            raise self.InvalidInputException(
+                'Invalid username: %s' % old_username)
+
+        if len(new_username) > constants.MAX_USERNAME_LENGTH:
+            raise self.InvalidInputException(
+                'Expected new username to be less than %s characters, '
+                'received %s' % (constants.MAX_USERNAME_LENGTH, new_username))
+
+        if user_services.is_username_taken(new_username):
+            raise self.InvalidInputException('Username already taken.')
+
+        user_services.set_username(user_id, new_username)
```

Comment:
How will the user whose name got changed know about this? I don't think it will affect the login flow, but don't you need to notify them that the name has changed and the reason behind the change?

Reply:
Just to confirm that I'll be managing this process manually since it affects a very small, limited number of users. I don't think it needs to be implemented in an automated way!